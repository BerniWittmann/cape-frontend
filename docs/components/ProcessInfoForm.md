# ProcessInfoForm

A Form to show and edit basic information about a process

## Props

<!-- @vuese:ProcessInfoForm:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|process|The Process Object|`Object`|`true`|-|
|isNewProcess|Option whether this process needs to be created or updated|`Boolean`|`true`|-|

<!-- @vuese:ProcessInfoForm:props:end -->


## Methods

<!-- @vuese:ProcessInfoForm:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|submit|Submits the form and returns result on successful validation|callback with data on success or undefined if invalid|
|setFormPristine|Sets the Form Pristine|-|

<!-- @vuese:ProcessInfoForm:methods:end -->


