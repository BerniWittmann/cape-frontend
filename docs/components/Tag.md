# Tag

A Component to view a single Tag

## Props

<!-- @vuese:Tag:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tag|The Tag object|`Object`|`true`|-|
|closable|Option whether the tag should be deleteable or not|`Boolean`|`false`|false|
|size|Size of the Tag|`'medium'` / `'small'` / `'mini'`|`false`|-|

<!-- @vuese:Tag:props:end -->


## Events

<!-- @vuese:Tag:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|close|Emitted when the close button is clicked|-|

<!-- @vuese:Tag:events:end -->


