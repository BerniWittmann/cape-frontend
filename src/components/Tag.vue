<template>
  <el-tag class="tag" type="info" :style="style" :closable="closable" :size="size" @close="emitClose">{{ tag.name }}</el-tag>
</template>

<script>
import { convertHexToRgba } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * A Component to view a single Tag
 */
export default {
  name: 'Tag',
  props: {
    // The Tag object
    tag: {
      type: Object,
      required: true
    },

    // Option whether the tag should be deleteable or not
    closable: {
      type: Boolean,
      default: false
    },

    // Size of the Tag
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String
    }
  },

  computed: {
    style() {
      if (!this.tag.color) return {}

      return {
        backgroundColor: convertHexToRgba(this.tag.color, 0.1),
        borderColor: convertHexToRgba(this.tag.color, 0.2),
        color: convertHexToRgba(this.tag.color, 1)
      }
    }
  },

  methods: {
    emitClose() {
      // Emitted when the close button is clicked
      this.$emit('close')
    }
  }
}

</script>

<style lang="scss">
.tag {
  margin-right: 5px;
  &:last-of-type {
    margin-right: 0;
  }

  .el-tag__close {
    color: inherit !important;
    &:hover {
      background-color: inherit !important;
      color: #fff;
    }
  }
}
</style>
