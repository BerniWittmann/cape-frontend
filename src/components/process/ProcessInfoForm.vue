<template>
  <el-card class="process-info">
    <h4>{{ $t('process.edit.info')}}</h4>
    <el-form ref="processForm" :model="process" label-position="top" :rules="rules">
      <el-form-item :label="$t('process.edit.name')" prop="name">
        <el-input v-model="process.name"></el-input>
      </el-form-item>
      <el-form-item :label="$t('process.edit.tags')" prop="tags">
        <tag v-for="tag in process.tags" :tag="tag" :key="tag.id" closable @close="removeTag(tag)"></tag>
        <el-select
                v-if="tagInputVisible"
                v-model="newTag"
                size="small"
                ref="tagSelect"
                @keyup.enter.native="addTag"
                @change="addTag"
                @blur="hideTagInput"
                filterable
        >
          <el-option
                  v-for="tag in availableTags"
                  :key="tag.id"
                  :label="tag.name"
                  :value="tag">
            <tag :tag="tag" size="mini"></tag>
          </el-option>
        </el-select>
        <el-button v-if="!tagInputVisible && availableTags.length > 0" class="button-new-tag" size="small" @click="showTagInput" icon="el-icon-plus">{{
          $t('process.edit.add_tag') }}
        </el-button>
      </el-form-item>
    </el-form>
  </el-card>
</template>

<script>
/* ============
 * Tag Settings Component
 * ============
 *
 * A component to edit the Tags
 */

import TagComponent from '@/components/Tag.vue'

export default {
  components: {
    Tag: TagComponent
  },

  props: {
    process: {
      type: Object,
      required: true
    }
  },

  data() {
    return {
      rules: {
        name: [
          { required: true, message: this.$t('process.edit.validation.name.required'), trigger: 'blur' }
        ]
      },
      newTag: undefined,
      tagInputVisible: false
    }
  },

  computed: {
    allTags() {
      return this.$store.state.tag.tags
    },

    usedTagIDs() {
      return this.process.tags.map((t) => t.id)
    },

    availableTags() {
      return this.allTags.filter(t => !this.usedTagIDs.includes(t.id))
    }
  },

  methods: {
    submit(cb) {
      this.$refs.processForm.validate((valid) => {
        if (valid) {
          return cb(this.process)
        } else {
          return cb()
        }
      })
    },

    setFormPristine() {
      this.$refs.processForm.resetFields()
    },

    removeTag(tag) {
      this.process.tags = this.process.tags.filter((t) => t.id !== tag.id)
    },

    addTag() {
      this.process.tags.push(this.newTag)
    },

    showTagInput() {
      this.tagInputVisible = true
      this.$nextTick(() => {
        this.$refs.tagSelect.focus()
      })
    },

    hideTagInput() {
      setTimeout(() => {
        this.tagInputVisible = false
        this.newTag = undefined
      }, 100)
    }
  }
}

</script>

<style scoped lang="scss">
.process-info {
  background-color: #EBEEF5;
  border-radius: 5px;
  padding: 30px 20px;
}

.button-new-tag {
  margin-left: 5px;
}
</style>
