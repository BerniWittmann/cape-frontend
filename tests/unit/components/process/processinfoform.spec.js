import moment from 'moment'
import { mount, config, TransitionStub } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'

import ProcessInfoForm from '@/components/process/ProcessInfoForm.vue'
import Tag from '@/components/Tag.vue'

describe('Components', () => {
  describe('ProcessInfoForm', () => {
    let store
    let propsData
    let cmp
    config.stubs.transition = false

    const date = moment(moment().utc())
    console.warn = jest.fn()

    beforeEach((done) => {
      jest.useRealTimers()
      store = {
        state: {
          tag: {
            tags: [{
              id: '42',
              name: 'First Tag',
              color: '#FF0000'
            }, {
              id: '43',
              name: 'Second Tag',
              color: '#FFFF00'
            }, {
              id: '44',
              name: 'Third Tag',
              color: '#0FFF00'
            }, {
              id: '45',
              name: 'Fourth Tag',
              color: '#00FFF0'
            }]
          }
        },
        dispatch: jest.fn().mockImplementation(() => ({
          then: (arg) => arg()
        }))
      }
      propsData = {
        process: {
          id: '1',
          name: 'My Process',
          description: 'My Description',
          createdAt: date.clone().subtract(14, 'days'),
          lastEditedAt: date.clone().subtract(2, 'days'),
          tags: [{
            id: '42',
            name: 'First Tag',
            color: '#FF0000'
          }, {
            id: '43',
            name: 'Second Tag',
            color: '#FFFF00'
          }]
        },
        isNewProcess: false
      }
      render(done)
    })

    function render(done) {
      cmp = mount(ProcessInfoForm, {
        i18n,
        propsData,
        mocks: {
          $store: store
        },
        sync: false
      })
      cmp.vm.$nextTick(() => {
        done()
      })
    }

    function renderWithoutSelect() {
      config.stubs.transition = TransitionStub
      cmp = mount(ProcessInfoForm, {
        i18n,
        propsData,
        stubs: {
          'el-select': '<div></div>'
        },
        mocks: {
          $store: store
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })
    it('renders the correct title', () => {
      expect(cmp.find('.title span').find('span').exists()).toBeTruthy()
      expect(cmp.find('.title span').find('span').text()).toEqual('process.edit.title')
    })
    it('renders an input field for the description', () => {
      const input = cmp.find('textarea')
      expect(input.exists()).toBeTruthy()
      input.setValue('My new Name')
      expect(cmp.vm.data.description).toEqual('My new Name')
    })
    it('renders the tags', () => {
      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(2)
      expect(tags.at(0).props('tag')).toEqual(propsData.process.tags[0])
      expect(tags.at(1).props('tag')).toEqual(propsData.process.tags[1])
    })
    it('allows to edit the input for the title', (done) => {
      // cmp.setMethods({ showInput: jest.fn() })
      const editTitleButton = cmp.find('button.black-color')
      expect(cmp.vm.nameInputVisible).not.toBeTruthy()
      expect(editTitleButton.exists).toBeTruthy()
      editTitleButton.trigger('click')
      expect(cmp.vm.nameInputVisible).toBeTruthy()
      // expect(cmp.vm.showInput).toHaveBeenCalled()
      cmp.vm.$nextTick(() => {
        expect(cmp.html()).toMatchSnapshot()
        expect(cmp.find('.el-input__inner').exists).toBeTruthy()
        done()
      })
    })
    it('hides the input after editing title', (done) => {
      const editTitleButton = cmp.find('button.black-color')
      expect(editTitleButton.exists).toBeTruthy()

      editTitleButton.trigger('click')
      cmp.vm.$nextTick(() => {
        expect(cmp.html()).toMatchSnapshot()
        const input = cmp.find('.el-input__inner')
        input.setValue('My new Process Name')
        expect(cmp.vm.data.name).toEqual('My new Process Name')
        expect(cmp.vm.nameInputVisible).toBeTruthy()
        input.trigger('keyup', { key: 'Enter' })
        expect(cmp.vm.nameInputVisible).not.toBeTruthy()
        done()
      })
    })
    it('checks the input to not be empty', (done) => {
      const editTitleButton = cmp.find('button.black-color')

      editTitleButton.trigger('click')
      cmp.vm.$nextTick(() => {
        expect(cmp.html()).toMatchSnapshot()
        const input = cmp.find('.el-input__inner')
        input.setValue('')
        input.trigger('keypress', { key: 'Enter' })
        expect(cmp.html()).toMatchSnapshot()
        expect(cmp.find('.el-input__inner').exists).toBeTruthy()
        done()
      })
    })
    describe('works in edit mode as expected', () => {
      it('renders correct title prefix', () => {
        expect(cmp.findAll('.title').at(1).exists()).toBeTruthy()
        expect(cmp.findAll('.title').at(1).text()).toEqual(propsData.process.name)
      })
      it('has back, reset and save button', () => {
        const buttons = cmp.findAll('button.right-space')
        expect(buttons.length).toEqual(3)

        expect(buttons.at(0).text()).toContain('back')
        expect(buttons.at(1).text()).toContain('reset')
        expect(buttons.at(2).text()).toContain('save')
      })
    })
    describe('works in new process as expected', () => {
      it('renders correct title prefix', (done) => {
        propsData.process.name = undefined
        propsData.isNewProcess = true
        render(() => {
          expect(cmp.html()).toMatchSnapshot()
          done()
        })
        expect(cmp.vm.nameInputVisible).toBeTruthy()
        expect(cmp.find('.title span').find('span').exists()).toBeTruthy()
        expect(cmp.find('.title span').find('span').text()).toEqual('process.add.title')
      })
      it('has back and save button but no reset button', (done) => {
        propsData.isNewProcess = true
        render(() => {
          expect(cmp.html()).toMatchSnapshot()
          done()
        })
        const buttons = cmp.findAll('button.right-space')
        expect(buttons.length).toEqual(2)

        expect(buttons.at(0).text()).toContain('back')
        expect(buttons.at(1).text()).toContain('save')
      })
    })
    describe('it can add a tag', () => {
      it('has a addTag Button', () => {
        const addButton = cmp.find('.tag-space')
        expect(addButton.exists()).toBeTruthy()
      })
      it('can shows the option on new tag button click', (done) => {
        cmp.findAll('.tag-space').at(1).trigger('click')
        cmp.vm.$nextTick(() => {
          expect(cmp.html()).toMatchSnapshot()
          const options = cmp.findAll('.el-select-dropdown li')
          expect(options.length).toEqual(2)
          expect(options.at(0).text()).toEqual(store.state.tag.tags[2].name)
          expect(options.at(1).text()).toEqual(store.state.tag.tags[3].name)
          done()
        })
      })
      it('adds the new tag on click', (done) => {
        cmp.findAll('.tag-space').at(1).trigger('click')
        cmp.vm.$nextTick(() => {
          cmp.find('.el-select-dropdown li').trigger('click')
          expect(propsData.process.tags.length).toEqual(3)
          expect(propsData.process.tags[2]).toEqual(store.state.tag.tags[2])
          done()
        })
      })
      it('adds a tag just once', () => {
        expect(cmp.vm.process.tags.length).toEqual(2)
        cmp.vm.newTag = cmp.vm.process.tags[0]
        cmp.vm.addTag()
        expect(cmp.vm.process.tags.length).toEqual(2)
      })
      it('closes the dropdown after adding A Tag', (done) => {
        cmp.vm.tagInputVisible = true
        cmp.vm.$nextTick(() => {
          jest.useFakeTimers()

          cmp.vm.hideTagInput()

          jest.runOnlyPendingTimers()

          cmp.vm.$nextTick(() => {
            expect(cmp.html()).toMatchSnapshot()
            done()
          })
        })
      })
    })
    it('can remove a tag', () => {
      renderWithoutSelect()
      const tag = cmp.findAll(Tag).at(1)
      tag.vm.$emit('close')

      expect(cmp.vm.data.tags.length).toEqual(1)
    })
    it('hides the add tag button if no tags are available', (done) => {
      store.state.tag.tags = [store.state.tag.tags[0], store.state.tag.tags[1]]
      render(() => {
        expect(cmp.html()).toMatchSnapshot()
        done()
      })
    })
    describe('has a function to validate the form', () => {
      it('has the function defined', () => {
        expect(cmp.vm.submit).toEqual(expect.any(Function))
      })
      it('submits the result if valid', () => {
        cmp.vm.$refs.processForm.validate = jest.fn().mockImplementation((cb) => {
          // eslint-disable-next-line standard/no-callback-literal
          cb(true)
        })
        cmp.vm.submit((result) => {
          expect(result).toEqual({
            'name': 'My Process',
            'description': 'My Description',
            'tags': [{
              'color': '#FF0000',
              'id': '42',
              'name': 'First Tag'
            }, {
              'color': '#FFFF00',
              'id': '43',
              'name': 'Second Tag'
            }]
          })
        })
      })
      it('does not submit if its not valid', () => {
        cmp.vm.$refs.processForm.validate = jest.fn().mockImplementation((cb) => {
          // eslint-disable-next-line standard/no-callback-literal
          cb(false)
        })
        cmp.vm.submit((result) => {
          expect(result).toEqual(undefined)
        })
      })
    })
    describe('has a function to reset the form validation state', () => {
      it('has the function defined', () => {
        expect(cmp.vm.setFormPristine).toEqual(expect.any(Function))
      })
      it('resets the Form', () => {
        cmp.vm.$refs.processForm.resetFields = jest.fn()
        cmp.vm.setFormPristine()
        expect(cmp.vm.$refs.processForm.resetFields).toHaveBeenCalled()
      })
    })
    it('handles a change from the external process', (done) => {
      cmp.setProps({
        process: {
          name: 'New Prop Name',
          description: 'My Description',
          tags: []
        }
      })
      cmp.vm.$nextTick(() => {
        expect(cmp.vm.data.name).toEqual('New Prop Name')
        expect(cmp.vm.data.tags).toEqual([])
        expect(cmp.vm.data.description).toEqual('My Description')
        done()
      })
    })
  })
})
