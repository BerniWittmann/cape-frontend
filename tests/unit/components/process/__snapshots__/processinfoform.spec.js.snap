// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ProcessInfoForm hides the add tag button if no tags are available 1`] = `
<div class="process-info">
  <h4>process.edit.info</h4>
  <form class="el-form el-form--label-top">
    <div class="el-form-item is-required"><label for="name" class="el-form-item__label">process.edit.name</label>
      <div class="el-form-item__content">
        <div class="el-input">
          <!----><input type="text" autocomplete="off" class="el-input__inner">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
    <div class="el-form-item"><label for="tags" class="el-form-item__label">process.edit.tags</label>
      <div class="el-form-item__content"><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);" name="el-zoom-in-center">First Tag<i class="el-tag__close el-icon-close"></i></span><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 255, 0, 0.1); color: rgb(255, 255, 0);" name="el-zoom-in-center">Second Tag<i class="el-tag__close el-icon-close"></i></span>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </form>
</div>
`;

exports[`Components ProcessInfoForm it can add a tag can shows the option on new tag button click 1`] = `
<div class="process-info">
  <h4>process.edit.info</h4>
  <form class="el-form el-form--label-top">
    <div class="el-form-item is-required"><label for="name" class="el-form-item__label">process.edit.name</label>
      <div class="el-form-item__content">
        <div class="el-input">
          <!----><input type="text" autocomplete="off" class="el-input__inner">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
    <div class="el-form-item"><label for="tags" class="el-form-item__label">process.edit.tags</label>
      <div class="el-form-item__content"><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);">First Tag<i class="el-tag__close el-icon-close"></i></span><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 255, 0, 0.1); color: rgb(255, 255, 0);">Second Tag<i class="el-tag__close el-icon-close"></i></span>
        <div class="el-select el-select--small">
          <!---->
          <div class="el-input el-input--small el-input--suffix">
            <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
            <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span>
            <!----></span>
            <!---->
          </div>
          <div class="el-select-dropdown el-popper" style="display: none;">
            <div class="el-scrollbar" style="">
              <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                <ul class="el-scrollbar__view el-select-dropdown__list">
                  <!---->
                  <li class="el-select-dropdown__item"><span class="tag el-tag el-tag--info el-tag--mini el-zoom-in-center-enter el-zoom-in-center-enter-active" style="background-color: rgba(15, 255, 0, 0.1); color: rgb(15, 255, 0);">Third Tag</span></li>
                  <li class="el-select-dropdown__item"><span class="tag el-tag el-tag--info el-tag--mini el-zoom-in-center-enter el-zoom-in-center-enter-active" style="background-color: rgba(0, 255, 240, 0.1); color: rgb(0, 255, 240);">Fourth Tag</span></li>
                </ul>
              </div>
              <div class="el-scrollbar__bar is-horizontal">
                <div class="el-scrollbar__thumb" style="width: 0px; transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
              </div>
              <div class="el-scrollbar__bar is-vertical">
                <div class="el-scrollbar__thumb" style="height: 0px; transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
              </div>
            </div>
            <!---->
          </div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
  </form>
</div>
`;

exports[`Components ProcessInfoForm it can add a tag closes the dropdown after adding A Tag 1`] = `
<div class="process-info">
  <h4>process.edit.info</h4>
  <form class="el-form el-form--label-top">
    <div class="el-form-item is-required"><label for="name" class="el-form-item__label">process.edit.name</label>
      <div class="el-form-item__content">
        <div class="el-input">
          <!----><input type="text" autocomplete="off" class="el-input__inner">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
    <div class="el-form-item"><label for="tags" class="el-form-item__label">process.edit.tags</label>
      <div class="el-form-item__content"><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);">First Tag<i class="el-tag__close el-icon-close"></i></span><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 255, 0, 0.1); color: rgb(255, 255, 0);">Second Tag<i class="el-tag__close el-icon-close"></i></span>
        <!----> <button type="button" class="el-button button-new-tag el-button--default el-button--small">
          <!----><i class="el-icon-plus"></i><span>process.edit.add_tag
      </span></button>
        <!---->
      </div>
    </div>
  </form>
</div>
`;

exports[`Components ProcessInfoForm renders 1`] = `
<div class="process-info">
  <h4>process.edit.info</h4>
  <form class="el-form el-form--label-top">
    <div class="el-form-item is-required"><label for="name" class="el-form-item__label">process.edit.name</label>
      <div class="el-form-item__content">
        <div class="el-input">
          <!----><input type="text" autocomplete="off" class="el-input__inner">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
    <div class="el-form-item"><label for="tags" class="el-form-item__label">process.edit.tags</label>
      <div class="el-form-item__content"><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 0, 0, 0.1); color: rgb(255, 0, 0);">First Tag<i class="el-tag__close el-icon-close"></i></span><span class="tag el-tag el-tag--info" style="background-color: rgba(255, 255, 0, 0.1); color: rgb(255, 255, 0);">Second Tag<i class="el-tag__close el-icon-close"></i></span>
        <!----> <button type="button" class="el-button button-new-tag el-button--default el-button--small">
          <!----><i class="el-icon-plus"></i><span>process.edit.add_tag
      </span></button>
        <!---->
      </div>
    </div>
  </form>
</div>
`;
